### DISCUSSION TITLE
[BUG] Inconsistent motions with multiple waveClasses

### DISCUSSION THREAD

**Discussion by mmacn42** [WEC-Sim/WEC-Sim#1365] [Q&A] [answered]
**Describe the bug**
I'm trying to use multiple waveClasses to simulate multi-spectral ocean conditions and seeing unrealistic WEC behaviors. To try and better understand, I ran this simple comparison between two test cases: 

_Test 1_
WEC-Sim RM3 example model
A single `regularCIC` wave, with height 2.0 meters and period 13 seconds

_Test 2_
WEC-Sim RM3 example model
Four waveClasses, each with `regularCIC` waves and one quarter of the wave height of _Test 1_ (0.5 meters). Same period.

**Workstation information (REQUIRED):**
Describe the workstation:
- Operating System: Windows 11 Pro
- MATLAB/Simulink Release: 2024a
- WEC-Sim Release: 6.1.1, working off of commit "1cdd4631"

**WEC-Sim file(s):**
I've uploaded the `wecSimInputFile` for each test case and the `userDefinedFunctions` I used to generate the plots I pasted below. The model I am running in this test case is the RM3 in the WEC-Sim examples folder.

[multiwavetests.zip](https://github.com/user-attachments/files/17683589/multiwavetests.zip)

**How to reproduce:**
Running each of the two `wecSimInputFile`s should reproduce the models and plots I've included.

**Expected behavior**
I expect the two models to have identical responses, as the superimposed waves are the same. However I see very different responses and forces on both bodies, shown below. I wonder if something is being double counted? Or if there is a mistake in my implementation please let me know.

**Screenshots:**
The two tests produce identical wave elevations but different body motions and forces.

_Wave elevations:_
![image](https://github.com/user-attachments/assets/3bc88e68-41f8-43f2-ab8b-1f660d222f39)

_Body motions:_
![image](https://github.com/user-attachments/assets/5fdb1210-03cb-4b15-8b6a-f331743aa0ee)
![image](https://github.com/user-attachments/assets/b592c07d-32fc-4d56-9f3c-c2ad59a278a7)

_Body forces:_
![image](https://github.com/user-attachments/assets/e03a0e23-db2c-40a0-8472-d3f291505bc4)
![image](https://github.com/user-attachments/assets/59d51ffe-1657-4fde-960e-52e2b1c9bd72)

**Comment by MShabara** [comment]
Hi @mmacn42,

The input files you attached are identical. Identical input files will produce identical results (assuming you used same .h5 and same .slx files).

I tested both input files you provided, using the geometry, hydroData, and .slx file from the examples folder. After running both input files, I found that they generated identical results. Unfortunately, I wasn’t able to replicate the issue you described.

Let me know if there’s anything specific you’d like me to check or if there are additional details that could help troubleshoot this further.

Best,

**Comment by mmacn42** [comment]
Very sorry, I attached the wrong Test 2 file. Here is the correct one.

[wecSimInputFile_test2.zip](https://github.com/user-attachments/files/17719199/wecSimInputFile_test2.zip)

**Comment by MShabara** [comment]
Hi @mmacn42 

The current version of WEC-Sim does not define irregular waves by using multiple regular wave instances. While I understand that this approach can work in other software (such as OpenFOAM v11), it is not applicable in WEC-Sim. To define irregular waves, you'll need to use the specific options available in the WEC-Sim documentation for creating a single irregular wave instance.

**Comment by mmacn42** [comment]
Hi @MShabara, thanks for your response! I'm not sure I quite understand however. Could you explain why this approach cannot work in wecSim? As I understand it, in the `WEC-Sim_Applications` repo the "Multiple_Wave_Headings" example allows for multiple spectra to be superimposed.

**Comment by kmruehl** [comment]
@MShabara can you please add a code snippet of initiallzing the `waveClass` from the input file for this discussion?

**Comment by DougSlocum** [comment]
Good morning, below are the two code snippets:
**wecSimInputFile_test1.m**
`% Regular Waves with CIC`
`waves = waveClass('regularCIC');          % Initialize Wave Class and Specify Type `
`waves.height = 2;                       % Wave Height [m]`
`waves.period = 13;                         % Wave Period [s]`

**wecSimInputFile_test2.m**
`waves(1) = waveClass('regularCIC');`
`waves(1).height = 0.5;`
`waves(1).period = 13;`

`waves(2) = waveClass('regularCIC');`
`waves(2).height = 0.5;`
`waves(2).period = 13;`

`waves(3) = waveClass('regularCIC');`
`waves(3).height = 0.5;`
`waves(3).period = 13;`

`waves(4) = waveClass('regularCIC');`
`waves(4).height = 0.5;`
`waves(4).period = 13;`

Under linear hydrodynamic approaches we would expect the cumulative forcing on the WEC bodies to be identical between the two cases, but as mmacn42 showed above, that's not the behavior we're observing. 

I'd also like to add that we have replicated this behavior using irregular waves as well. If those files would be helpful we can post those too. 

Thanks for your help!

**Comment by MShabara** [comment]
Hi @DougSlocum 

Thanks for your message!

Currently, WEC-Sim does not support defining irregular waves using multiple regular waves. This limitation arises because initializeWecSim.m only calculates the excitation coefficients for the first wave instance. It doesn’t loop through additional regular wave instances, as illustrated in the figure below:

![image](https://github.com/user-attachments/assets/6a13ae38-4d9d-491f-a11c-1bcef9832ecc)

To enable this functionality in WEC-Sim, several modifications would be required, particularly in the bodyClass.m file and the body simulink library. We've added this enhancement to our to-do list for future releases.

In the meantime, you can do this instead:
      1. Use WEC-Sim to create the combined waveform you need.
      2. Save the combined wave shape as a .mat file.
      3. Import that file into a new WEC-Sim run using the code below:

`% Waves with imported wave elevation time-history`  
`waves = waveClass('elevationImport');          % Create the Wave Variable and Specify Type `
`waves.elevationFile = 'elevationData.mat';     % Name of User-Defined Time-Series File [:,2] = [time, eta]  `

This method allows you to directly incorporate any wave elevation time series into WEC-Sim.

Let me know if you have any questions.

**Comment by DougSlocum** [comment]
Hi @MShabara, thanks for your response. I'm afraid to say I'm confused on a few points.

1.  We are not trying to define irregular waves using regular waves, rather we are trying to use the regular waves in this example as a test case to communicate the odd behavior.
2. In the code snippet you provided, it appears that the case with multiple waves super-imposed upon one another should only be using 'waves(1).height = 0.5', which is then driving a much higher forcing than just 'waves.height = 2', which I don't quite follow.
3. How does WEC-Sim evaluate wave conditions on the body like the example shown here? Is it only using 'waves(1)' in this case? [https://github.com/WEC-Sim/WEC-Sim_Applications/tree/main/Multiple_Wave_Headings]
4. We also get similar results (i.e. different forcing on bodies) using a spectrum definition. We ran these two conditions on the RM3 model.
  **Single spectrum**
  ```
  waves.height = 02;
  waves.period = 13;
  waves.spectrumType = 'PM';
  waves.phaseSeed = 2;
```
**Multiple spectra that superimpose to be equal to the single spectrum**
```
  waves(1).height = 01;
  waves(1).period = 13;
  waves(1).spectrumType = 'PM';
  waves(1).phaseSeed = 2;
  waves(2).height = 01;
  waves(2).period = 13;
  waves(2).spectrumType = 'PM';
  waves(2).phaseSeed = 2;
```
These result in the following differences in forcing. 
![image](https://github.com/user-attachments/assets/8f360cd8-8166-4406-a066-8fd08aa09206)

5. The TEAMER WEC workshop strongly emphasized the importance of considering multi-directional sea states in design, which is why we're digging at this. When we ran a sea state with only 1 wave class, it behaved as expected. When we ran any simulation with two or more wave classes, our WEC behaved erratically, including coming entirely out of the water, which is what led us to start digging at this.

**Reply by DougSlocum** [reply]
I realized I forgot to include a plot of wave elevation to accompany the body forces for our irregular spectrum case, sorry about that. 
![image](https://github.com/user-attachments/assets/69d2a8fb-8d13-4983-8096-40c7ae5b0d4e)

**Comment by MShabara** [comment]
Hi @mmacn42 and @DougSlocum 

Thank you for bringing this issue to our attention! Thanks to your comments we've identified a bug in the latest WEC-Sim release (v6.1.x), and our team is actively working on a fix in PR #1373 . This will be resolved within the next two weeks.

The main differences between versions 6.1.x and 6.0 are related to variable hydrodynamics and QTF calculations. If your model does not require these features, we recommend using WEC-Sim v6.0 as a temporary workaround until the PR is reviewed and merged within the main repo.

Let me know if you have any questions.

**Comment by DougSlocum** [comment]
@MShabara, thanks for the update!

**Comment by MShabara** [comment, marked as answer]
@DougSlocum @mmacn42 The bug was solved in the main branch.
